packages/@ember/-internals/meta/lib/meta.ts

```diff
@@ -449,82 +493,211 @@ export class Meta {
     method: Function | string,
     once: boolean
   ) {
-    if (this._listeners === undefined) {
-      this._listeners = [];
+    if (DEBUG) {
+      counters!.addToListenersCalls++;
     }
-    this._listeners.push(eventName, target, method, once);
+
+    this.pushListener(eventName, target, method, once ? ListenerKind.ONCE : ListenerKind.ADD);
   }
 
-  _finalizeListeners() {
-    if (this._listenersFinalized) {
-      return;
+  removeFromListeners(eventName: string, target: object | null, method: Function | string): void {
+    if (DEBUG) {
+      counters!.removeFromListenersCalls++;
     }
-    if (this._listeners === undefined) {
-      this._listeners = [];
+
+    this.pushListener(eventName, target, method, ListenerKind.REMOVE);
+  }
+
+  removeAllListeners(event: string) {
+    if (DEBUG) {
+      counters!.removeAllListenersCalls++;
     }
-    let pointer = this.parent;
-    while (pointer !== null) {
-      let listeners = pointer._listeners;
-      if (listeners !== undefined) {
-        this._listeners = this._listeners.concat(listeners);
+
+    let listeners = this.writableListeners();
+    let inheritedEnd = this._inheritedEnd;
+    // remove all listeners of event name
+    // adjusting the inheritedEnd if listener is below it
+    for (let i = listeners.length - 1; i >= 0; i--) {
+      let listener = listeners[i];
+      if (listener.event === event) {
+        listeners.splice(i, 1);
+        if (i < inheritedEnd) {
+          inheritedEnd--;
+        }
       }
-      if (pointer._listenersFinalized) {
-        break;
+    }
+    this._inheritedEnd = inheritedEnd;
+    // we put remove alls at start because rare and easy to check there
+    listeners.splice(inheritedEnd, 0, {
+      event,
+      target: null,
+      method: null,
+      kind: ListenerKind.REMOVE_ALL,
+    });
+  }
+
+  private pushListener(
```

Comment:
Should we have a debug counter for this one?

Reply:
`pushListener` exists just to deduplicate some code from `removeFromListeners` and `addToListeners`, and we have counters for both of those. It's also important to be able to tell what kind of listener object we are pushing (e.g. so we can tell if removing is much more rare than adding), so having unique counters per-wrapper function makes more sense here I think.